### PR TITLE
fix(analysis): align pragma and scope diagnostics (#3339)

### LINKED ISSUE #3339: use v5.40; incorrectly triggers "add use strict; use warnings;" suggestions
## Bug Description

When a Perl file declares `use v5.40;` (or any version ≥5.12), the LSP still suggests adding `use strict;` and `use warnings;` even though Perl's `use VERSION` pragma implicitly enables strict (5.12+) and warnings (5.35+).

## Expected Behavior

Files with `use v5.12;` or higher should not trigger "missing strict" diagnostics.
Files with `use v5.35;` or higher should not trigger "missing warnings" diagnostics.

## Actual Behavior

The following code triggers both PL100 (missing-strict) and PL101 (missing-warnings) diagnostics:

```perl
use v5.40;
my $x = 1;
```

## Root Cause

Three separate views of "effective Perl semantics" exist in the codebase without coordination:

1. **`strict_warnings.rs`** — Only recognizes literal `use strict`, `use warnings`, or modules like `Modern::Perl`, `Moo`, `Moose`, `Mojo::Base`. It does not check version declarations.

2. **`version_compat.rs`** — Parses `use vN.N` and derives feature sets (say, state, signatures, etc.), but does not consult version for strict/warnings implication.

3. **`PragmaTracker` (perl-pragma)** — Only tracks explicit `use/no strict` and `use/no warnings`, not version-implied pragmas.

## Perl Documentation References

Per [perlpragma](https://perldoc.perl.org/perlpragma) and [use](https://perldoc.perl.org/functions/use):

> The `use VERSION` form... if VERSION is 5.12.0 or higher, strictures are enabled lexically... if VERSION is 5.35.0 or higher, warnings are enabled.

## Suggested Fix

Unify version/pragma semantics in a shared layer:

- Extend `PragmaTracker` to recognize `use VERSION` declarations and set effective strict/warnings state accordingly
- OR have `check_strict_warnings()` query the version_compat layer to determine if strict/warnings are implicitly enabled
- Consider a unified "effective pragma state" that both systems consult

## Files to Reference

- `crates/perl-lsp-diagnostics/src/lints/strict_warnings.rs`
- `crates/perl-lsp-diagnostics/src/lints/version_compat.rs`
- `crates/perl-pragma/src/lib.rs` (PragmaTracker)

### PR DESCRIPTION
**Pull request by EffortlessSteven**
## Summary
- centralize lexical `use VERSION` semantics in `perl-pragma` and reuse them across strict/warnings checks and version compatibility (#3339)
- split scope analysis so undeclared variables follow `strict 'vars'` while barewords follow `strict 'subs'` (#3339)
- normalize dereference/container use tracking so `@$arrayref`, `@{$arrayref}`, and `$hashref->{k}` mark the right declarations as used (#3338)
- treat declaration-capable builtin handle arguments as consumed and initialized after calls like `open my $fh, ...` (#3336)

## Commentary Takeaways
- #3336 commentary was right to question parser vs analyzer, but the parser AST was already correct; the fix belongs in scope consumption.
- #3338 commentary found the right file, but the safe fix is syntax-aware normalization, not blanket cross-sigil aliasing.
- #3339 is the shared-semantics seam and leads the implementation order here.

## Validation
- `cargo fmt --all`
- `cargo test -p perl-pragma`
- `cargo test -p perl-semantic-analyzer --test scope_and_symbol_tests -- --nocapture`
- `cargo test -p perl-lsp-diagnostics --test version_compat_tests --test diagnostic_integration_test -- --nocapture`
- `cargo test -p perl-lsp-diagnostics --lib -- --nocapture`
- `just pr-fast`

Closes #3339
Related: #3336, #3338

### CODE REVIEW

**Comment by EffortlessSteven** [comment]
## PR Review: Align Pragma and Scope Diagnostics

### Summary Assessment

This PR addresses **three issues**: #3336 (builtin declaration initializers), #3339 (version pragma implied strict/warnings), and #3338 (dereference use tracking). The implementation is **directionally correct** and fixes the reported bugs. Below is a technical comparison against the architectural recommendations from issue commentary.

---

### #3339: Version Pragma Handling — Partial Alignment with Recommendations

**What We Recommended:**
- A unified `EffectiveSemantics` layer combining explicit pragmas + version-derived + OO framework sources
- OR extending `PragmaTracker` with version awareness

**What the PR Implements:**
- **Took the PragmaTracker extension approach** — this matches our "Option A" recommendation

Key implementation choices:
- Added `PerlVersion` struct + centralized `parse_perl_version()` in `perl-pragma`
- Added `version_implies_strict()` / `version_implies_warnings()` threshold functions
- Modified `PragmaTracker` to recognize `use v5.x` and set effective strict/warnings state
- Updated `strict_warnings.rs` to query `PragmaTracker` instead of doing literal matching

**Assessment:** This is a **pragmatic, correct implementation**. It does not go full `EffectiveSemantics` (which would add provenance tracking and source-of-truth metadata), but it solves the immediate bug with significantly less code. The trade-off is acceptable.

**One divergence to note:** The oppositional review in #3339 suggested a targeted fix directly in `strict_warnings.rs` (~15 lines) to avoid touching `PragmaTracker`. This PR takes the slightly heavier but architecturally cleaner route of extending `PragmaTracker`. **Good call** — the targeted fix would have created technical debt; this consolidates pragma knowledge where it belongs.

---

### #3336: Builtin Initializers — Simpler Than Recommended

**What We Recommended:**
- A systematic `BUILTIN_INITIALIZERS` PHF registry in `perl-builtins-phf` alongside `BUILTIN_SIGS`
- ~150 lines including registry definition + scope analyzer integration + comprehensive tests
- Covers: `open`, `sysopen`, `opendir`, `pipe`, `socket`, `socketpair`, `accept`, `read`, `sysread`, `recv`, `dbmopen`

**What the PR Implements:**
- A hardcoded `is_declaration_capable_builtin()` function matching: `open`, `opendir`, `sysopen`, `pipe`, `socket`, `accept`
- `mark_builtin_declaration_arg_consumed()` helper that finds `VariableDeclaration` nodes in builtin args and marks them initialized
- ~30 lines of implementation in `scope_analyzer.rs`

**What is Missing vs. Recommendation:**
- No PHF registry in `perl-builtins-phf` — the list is hardcoded in `scope_analyzer.rs`
- Only covers 6 builtins vs. the 11 we identified
- Missing: `socketpair`, `read`, `sysread`, `recv`, `dbmopen`
- Position-aware handling (`read FH, my $buf` initializes position 1, not 0)

**Assessment:** This is a **minimal viable fix** for the most common case (`open my $fh`). It solves the reported bug but creates **future work**:
- Users will eventually hit false positives on `pipe my $r, my $w` (second arg), `read FH, my $buf`, etc.
- The hardcoded list will need to be manually kept in sync with Perl's evolving builtin set

**Suggestion:** Consider a follow-up PR to extract this into the `BUILTIN_INITIALIZERS` registry pattern. The infrastructure is simple (~30 lines for the PHF map) and the payoff is eliminating a class of similar bugs.

---

### #3338: Dereference Use Tracking — Good Syntax-Aware Fix

The PR includes a `resolve_variable_use_target()` helper that normalizes dereference syntax:
- `@$ref` counts as use of `$ref`
- `$arr[0]` counts as use of `@arr`
- `$hash{k}` counts as use of `%hash`

This appears to correctly address the cross-sigil aliasing issue we identified in #3338 commentary. The implementation is syntax-aware rather than blanket-aliasing, which is the right call.

---

### #3339: Strict/Subs Split — Correct Fix

The PR splits `strict_mode` into:
- `strict_vars_mode` controls undeclared variable diagnostics (PL103)
- `strict_subs_mode` controls bareword diagnostics (PL105)

This matches Perl's actual `strict 'vars'` / `strict 'subs'` distinction and fixes the over-reporting issue noted in #3339.

---

### Test Coverage

**Good coverage for the implemented features:**
- `version_pragma_suppresses_strict_warnings_diagnostic` — v5.40 implies both
- `v5_12_suppresses_strict_but_not_missing_warnings` — threshold correctness
- `strict_vars_only_checks_undeclared_variables` / `strict_subs_only_checks_barewords` — split verification
- `scope_declaration_capable_builtins_initialize_handle_bindings` — builtin initializers
- `unused_variable_used_via_explicit_dereference_forms` — dereference tracking

**Missing tests for edge cases we identified:**
- `pipe my $r, my $w` with second arg usage
- `open(my $fh, ...)` parenthesized form
- `open my $fh, ... or die` flow-control form
- `no strict 'refs'` after `use v5.40` override
- Dev versions: `use 5.012_001`
- Three-part v-strings: `use v5.40.1`

---

### Code Quality Notes

1. **Version parsing robustness:** The `parse_perl_version()` implementation handles v-strings and decimal forms. Consider adding a test for underscore dev versions (`5.012_001`) which Perl treats as 5.12 but the current parser might reject.

2. **Multi-position initializers:** The current `mark_builtin_declaration_arg_consumed()` recursively processes all args for declaration-capable builtins. For `pipe my $r, my $w`, this should work since both args are declarations. For future `read FH, my $buf`, this will incorrectly try to initialize `FH` — a position-aware check would be safer.

3. **Performance:** The `PragmaTracker::build()` call in `strict_warnings.rs` adds an AST walk where there wasn't one before. For large files this is O(n) overhead. The trade-off is acceptable for correctness, but worth monitoring.

---

### Overall Verdict

**Approve with minor suggestions.**

This PR:
- Correctly fixes the reported bugs
- Takes a pragmatic approach that balances correctness vs. implementation complexity
- Centralizes version-derived pragma logic in `perl-pragma` (good architectural direction)
- Addresses the related #3338 dereference issue
- Leaves room for future enhancement (BUILTIN_INITIALIZERS registry, more complete test coverage)

**Recommendations:**
1. Land this PR — it solves the immediate user pain
2. Follow-up: Extract builtin initializers to a PHF registry and add position-aware handling
3. Follow-up: Add edge case tests (dev versions, `no strict` overrides, parenthesized forms)

The commentary from #3336 and #3339 served its purpose: the final implementation is cleaner than an ad-hoc targeted fix would have been, even if it didn't go all the way to the "proper" registry architecture. That is a win for oppositional review.

---

*References: #3336, #3339, #3338*

**Comment by EffortlessSteven** [comment]
## Follow-up: Addressing Gaps & Improvements

Based on the review, here are specific recommendations for improving this PR before or after merge:

### #3336 Improvements (Builtin Initializers)

**Current:** Hardcoded  with 6 builtins.

**Recommended path to full registry:**
1. **Immediate (this PR or fast-follow):** Add the missing 5 builtins:
   -  (dual initializer like )
   - , ,  (position 1, not 0)
   -  (single position 0)

2. **Future PR:** Migrate to  PHF map in :
   
   This mirrors existing  infrastructure and eliminates string matching in hot path.

### #3339 Improvements (Version→Pragma)

**Current:** Extended  with  struct.

**Test gaps to add:**
- Dev versions: , 
- Three-part v-strings: 
- Override semantics:  (should NOT warn)
- Boundary edge:  (just before strict) vs  (strict enabled)

### #3338 Improvements (Dereference)

**Current:**  handles sigil bridging.

**Test gaps to add:**
-  second arg initialization: 
- Parenthesized forms: 
- Complex derefs: , 

### Suggested PR Structure

1. **Merge this PR as-is** — it solves immediate user pain
2. **Fast-follow PR:** Add missing builtins + test coverage
3. **Future PR:** PHF registry refactor when touching builtins crate anyway

The architectural research we did (unified pragma layer, PHF registry) can be referenced when someone has bandwidth for the deeper refactor.

cc: @EffortlessSteven

**Comment by EffortlessSteven** [comment]
## Follow-up: Addressing Gaps & Improvements

Based on the review, here are specific recommendations for improving this PR before or after merge:

### #3336 Improvements (Builtin Initializers)

**Current:** Hardcoded `is_declaration_capable_builtin()` with 6 builtins.

**Recommended path to full registry:**

1. **Immediate (this PR or fast-follow):** Add the missing 5 builtins:
   - `socketpair` (dual initializer like `pipe`)
   - `read`, `sysread`, `recv` (position 1, not 0)
   - `dbmopen` (single position 0)

2. **Future PR:** Migrate to `BUILTIN_INITIALIZERS` PHF map in `perl-builtins-phf`:
   ```rust
   pub static BUILTIN_INITIALIZERS: phf::Map<&'static str, &[usize]> = phf::phf_map! {
       "open" => &[0],
       "pipe" => &[0, 1],
       "read" => &[1],  // position 1, not 0!
       // ... etc
   };
   ```
   This mirrors existing `BUILTIN_SIGS` infrastructure and eliminates string matching in hot path.

### #3339 Improvements (Version→Pragma)

**Current:** Extended `PragmaTracker` with `PerlVersion` struct.

**Test gaps to add:**
- Dev versions: `use v5.12.1;`, `use 5.012_001;`
- Three-part v-strings: `use v5.40.1;`
- Override semantics: `use v5.40; no strict; my $x = 1;` (should NOT warn)
- Boundary edge: `use v5.11;` (just before strict) vs `use v5.12;` (strict enabled)

### #3338 Improvements (Dereference)

**Current:** `resolve_variable_use_target()` handles sigil bridging.

**Test gaps to add:**
- `pipe` second arg initialization: `pipe my $r, my $w;`
- Parenthesized forms: `open(my $fh, '<', 'file');`
- Complex derefs: `@{ $ref }`, `@$obj->{field}`

### Suggested PR Structure

1. **Merge this PR as-is** — it solves immediate user pain
2. **Fast-follow PR:** Add missing builtins + test coverage
3. **Future PR:** PHF registry refactor when touching builtins crate anyway

The architectural research we did (unified pragma layer, PHF registry) can be referenced when someone has bandwidth for the deeper refactor.

File: crates/perl-semantic-analyzer/src/analysis/scope_analyzer.rs [hunk 1]
```diff
@@ -932,6 +915,41 @@ impl ScopeAnalyzer {
...
+    fn resolve_variable_use_target<'a>(
+        &self,
+        node: &'a Node,
+        ancestors: &[&'a Node],
+    ) -> Option<(&'a str, &'a str)> {
+        let NodeKind::Variable { sigil, name } = &node.kind else {
+            return None;
+        };
+
+        if (sigil == "@" || sigil == "%") && name.starts_with('$') && name.len() > 1 {
```
[the remark this answers is not in the document]
Comment (EffortlessSteven) [reply crates/perl-semantic-analyzer/src/analysis/scope_analyzer.rs]: Addressed in `c0728dff`. `resolve_variable_use_target` now normalizes scalar dereferences like `$$ref` back to the declared `$ref`, and I added regressions covering both unused-variable tracking and strict-mode resolution in `crates/perl-semantic-analyzer/tests/scope_and_symbol_tests.rs`.
